File: server/controllers/UserController.js
const User = require('../model/User');
const cloudinary = require('cloudinary').v2;

// Configure Cloudinary once at module load
cloudinary.config({
  cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
  api_key: process.env.CLOUDINARY_API_KEY,
  api_secret: process.env.CLOUDINARY_API_SECRET,
});

class UserController {
  static async getMe(req, res) {
    try {
      const auth = req.auth || {};
      if (auth.type !== 'user' || !auth.userId) {
        return res.status(401).json({ message: 'Unauthorized' });
      }

      const user = await User.findById(auth.userId).lean();
      if (!user) return res.status(404).json({ message: 'User not found' });

      return res.json({
        user: {
          id: user._id,
          email: user.email || '',
          name: user.name || '',
          phone: user.phone || '',
          location: user.location || '',
          // Prefer avatar, fallback to picture
          avatar: user.avatar || user.picture || '',
          isGoogle: !!user.googleId,
        },
      });
    } catch (err) {
      return res.status(500).json({ message: err.message });
    }
  }

  static async updateMe(req, res) {
    try {
      const auth = req.auth || {};
      if (auth.type !== 'user' || !auth.userId) {
        return res.status(401).json({ message: 'Unauthorized' });
      }

      const { name = '', phone = '', location = '', avatar = '' } = req.body || {};
      const updated = await User.findByIdAndUpdate(
        auth.userId,
        {
          name,
          phone,
          location,
          avatar,
        },
        { new: true }
      ).lean();

      if (!updated) return res.status(404).json({ message: 'User not found' });

      return res.json({
        user: {
          id: updated._id,
          email: updated.email || '',
          name: updated.name || '',
          phone: updated.phone || '',
          location: updated.location || '',
          avatar: updated.avatar || updated.picture || '',
        },
        message: 'Profile updated',
      });
    } catch (err) {
      return res.status(500).json({ message: err.message });
    }
  }

  static async uploadAvatar(req, res) {
    try {
      const auth = req.auth || {};
      if (auth.type !== 'user' || !auth.userId) {
        return res.status(401).json({ message: 'Unauthorized' });
      }

      const { image } = req.body || {};
      if (!image || typeof image !== 'string') {
        return res.status(400).json({ message: 'Missing image' });
      }

      const folder = process.env.CLOUDINARY_FOLDER_PROFILE || 'profile-avatar';
      const result = await cloudinary.uploader.upload(image, {
        folder,
        resource_type: 'image',
        overwrite: true,
      });

      return res.status(200).json({ url: result.secure_url });
    } catch (err) {
      return res.status(500).json({ message: err.message });
    }
  }
}

module.exports = UserController;

File: server_2/controllers/user_controller.py
import os
from datetime import datetime
from typing import Any, Dict

import cloudinary
import cloudinary.uploader
from bson import ObjectId
from fastapi import HTTPException, status
from motor.motor_asyncio import AsyncIOMotorDatabase
from pymongo import ReturnDocument

cloudinary.config(
    cloud_name=os.getenv("CLOUDINARY_CLOUD_NAME"),
    api_key=os.getenv("CLOUDINARY_API_KEY"),
    api_secret=os.getenv("CLOUDINARY_API_SECRET"),
)


def _ensure_object_id(value: str) -> ObjectId:
    try:
        return ObjectId(value)
    except Exception as exc:  # pylint: disable=broad-except
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid user id") from exc


def serialize_user(user: Dict[str, Any]) -> Dict[str, Any]:
    serialized = {
        "id": str(user.get("_id")),
        "email": user.get("email", ""),
        "name": user.get("name", ""),
        "phone": user.get("phone", ""),
        "location": user.get("location", ""),
        "avatar": user.get("avatar") or user.get("picture") or "",
        "isGoogle": bool(user.get("googleId")),
    }
    return serialized


async def get_me(db: AsyncIOMotorDatabase, user_id: str) -> Dict[str, Any]:
    user = await db.users.find_one({"_id": _ensure_object_id(user_id)})
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")
    return {"user": serialize_user(user)}


async def update_me(db: AsyncIOMotorDatabase, user_id: str, payload: Dict[str, Any]) -> Dict[str, Any]:
    updates = {
        "name": payload.get("name", ""),
        "phone": payload.get("phone", ""),
        "location": payload.get("location", ""),
        "avatar": payload.get("avatar", ""),
        "updatedAt": datetime.utcnow(),
    }

    updated = await db.users.find_one_and_update(
        {"_id": _ensure_object_id(user_id)},
        {"$set": updates},
        return_document=ReturnDocument.AFTER,
    )
    if not updated:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found")

    return {"user": serialize_user(updated), "message": "Profile updated"}


async def upload_avatar(image: str) -> str:
    if not image or not isinstance(image, str):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Missing image")

    folder = os.getenv("CLOUDINARY_FOLDER_PROFILE", "profile-avatar")
    result = cloudinary.uploader.upload(
        image,
        folder=folder,
        resource_type="image",
        overwrite=True,
    )
    return result.get("secure_url") or ""


async def delete_me(db: AsyncIOMotorDatabase, user_id: str) -> Dict[str, str]:
    oid = _ensure_object_id(user_id)
    await db.users.find_one_and_delete({"_id": oid})
    await db.resumes.delete_many({"userId": oid})
    return {"message": "Account and related data deleted"}
